### DISCUSSION TITLE
Long polling silently hangs or swallows errors

### DISCUSSION THREAD

**Discussion by wojpawlik** [telegraf/telegraf#1234] [Ideas] [answered]
1. If we don't await the middleware before fetching further updates, we break
   a) any semblance of sensible update confirming,
   b) [error handling](https://github.com/telegraf/telegraf/blob/bc18691c5c80a27796ef318d65ec431b7280d713/src/core/network/polling.ts#L74-L79) (every error turns into a rejection not handled by Telegraf),
   c) backpressure (bot will fetch updates even when overloaded),
   d) others, see https://github.com/telegraf/telegraf/issues/892#issuecomment-583765906.

2. If we do naively `await` (default behavior before #1224), bot will silently wait forever for handlers that break by never resolving (#735). This behavior is still available when `handlerTimeout === Infinity`, accessing it should probably [emit a warning](https://nodejs.org/api/process.html#process_process_emitwarning_warning_type_code_ctor).

3. In #1224, I changed the default behavior to resolve the promise if processing takes longer than 5 seconds. When phrased this way, the problem is obvious: errors thrown after 5 seconds are swallowed!

I have a suggestion, but it ain't pretty either: **throw** when processing takes longer than `handlerTimeout` (which'd need to be extended to say 30 seconds). Such a drastic change obviously needs to be decided upon before 4.0.

The real solution here would be to fetch updates via WebSocket instead, which could allow confirming on per-update basis.

In the meantime, please use webhook in production if you can (with `webhookReply` disabled: #1082).

**Comment by yeldiRium** [comment]
Why resolve the promise after a timeout at all?

If a handler does not resolve, that is a bug on the side of the bot developer. I don't think telegraf should try to recover from bugs made by a bot developer.

If the handler resolves (or errors) after a long time, the timeout swallows its results.

I'm sorry to interject since I don't know the history of some linked issues. But the swallowing of errors seems to me to be the result of trying to fix other's mistakes which results in more room for error. Or am I missing something?

**Reply by wojpawlik** [reply]
I'm just trying to fix #735. Please read it, it's all the context you need. Stopping processing updates without even a log message is unacceptable.

It's easy to say what _shouldn't_ happen in this case. What do you think _should_ happen instead?

Of course, Telegraf can't fix bugs in consumers' code, but it can do its best to inform about the issue and perhaps try to recover.

**Comment by KnorpelSenf** [comment]
I'd say you're right about this. When the middleware simply does not return, we cannot do much about it. It's upon the user to handle such a scenario, and that's what we should make possible.

The thing I have in mind is to let the user provide a `timeoutHandler`, i.e. write something like `bot.catchTimeout`. This handler will be supplied with the respective context object that lead to the timeout, and maybe the unresolved `Promise` (not sure about that one yet).

**Comment by KnorpelSenf** [comment]
Here we go.

---

We can solve the problem by introducing a new queue module that tracks all running promises. This is done as follows.

## What we can do about this

The central element of this suggestion is to use a list of promises. Each promise corresponds to one update that was passed to the middleware stack.

When a new update is incoming, the context object is created from it (that's what currently happens in `handleUpdate`). The context object can then be passed to the queue module, which will turn it into a promise by applying it to the middleware stack. The promise is added to the list.

Once the promise resolves or rejects, it is removed from the list again. If it rejects, an error handler is invoked (think: the current `catch`) with both the error and the corresponding context object that led to the error.

If the promise is neither resolved nor rejected before the timeout, it is removed from the promise queue anyway. It can be passed to a second error handler along with its context object. It is upon the user to handle this, e.g. via `catchTimeout`.

## Why this is going to work

Going through the problems mentioned in the issue description above, I'm confident that we can solve every single problem there.

- we do await the middleware, so 1) does not apply
- we do not do this naïvely, so 2) does not apply (the bot won't wait forever for middleware)
- no errors are ever swallowed, so 3) is no longer an issue
- we have a clean way to handle promises that time out
- we do not need websocket support
- we can still track success and failure on a per-update basis
- this works with long polling and webhooks, hence bringing the error handling benefits to both modi

Also, we will basically get a few nice things for free.

- much better performance because we parallelize middleware execution and update polling, so we no longer have to wait for the middleware stack to complete before we can even begin fetching new updates
- basic backpressure support because we can determine the number of running tasks at any point, so the polling mechanism can adjust accordingly and automatically

## What problems we have to solve here

1) We cannot determine whether a promise is actually already resolved or not, at least not as of 2020.
2) We have to find a good data structure. It must support adding an element in `O(1)`. It must support removing an element in an arbitrary position in `O(1)`.
3) It must be possible to determine which promises timed out, and remove them, and this should be in `O(n)` where `n` is the number of promises to remove.
4) We should avoid starting a timeout for every single incoming update. We should rely on a single timeout can handle all updates.
5) It should be able to automatically pause the polling if a concurrency limit is reached, and the polling should continue afterwards. This should be possible with just minimal changes to the polling module because we don't want to add `pause` and `resume` flags and interrupts etc.

## How we can achieve all of that

I have a nice solution for all of the above, but it makes sense to define two important terms first.

### “Decaying Deque”

A _Decaying Deque_ is a special form of doubly linked list. It has a timestamp for each node, and it is guaranteed to be sorted by these timestamps at all times.

It keeps track of its length at all times.

Usually, deques allow elements to be removed from both sides, but we will rather use it to remove elements from one side, or at arbitrary positions.

It is called decaying deque because you can think of it like we're adding updates to one side, but arbitrary elements are picked out at any time, hence performing a dequeue operation from the centre. The further left an update is, the higher its chance to be removed. Its elements _decay_.

### “Drift”

Once we turn an element (i.e. a context object) into a promise, we don't actually add this to the decaying deque. Instead, we transform it further to a drift. You can think of a drift as a self-removing element of a decaying deque, in other words, just a special form of node in a special form of a linked list.

The intuition behind the word drift is that the promise is _drifting_ to the left in the decaying deque, until it suddenly removes itself as soon as its promise completes.

### Everything in operation

A `getUpdates` call fetches 10 updates. They are turned into drifts and added to the tail of the decaying deque. This can be done linearly in the number of updates (10).

```
Previous list:
[]

Next list:
[0,1,2,3,4,5,6,7,8,9]
```

A timeout pointer is set to the _tail_ of the decaying deque. We know that all 10 updates were added in the same instant, so we only need to store one timestamp, remember one element, and start one timer.

```
head          tail   timeout pointer
 |                \ /
 V                 V
[0,1,2,3,4,5,6,7,8,9]
```

When the next batch of updates comes in, it is appended to the decaying deque without moving the timeout pointer.

```
            timeout pointer
                   |
head               |    tail
 |                 |     |
 V                 V     V
[0,1,2,3,4,5,6,7,8,9,a,b,c]
```

If a promise resolves or rejects, it can remove itself from the queue. If the promise resolves that one of the pointers is pointing at, it will have to adjust the respective pointer accordingly. Here, 9 resolved:

```
          timeout pointer
                 |
head             |    tail
 |               |     |
 V               V     V
[0,1,2,3,4,5,6,7,8,a,b,c]
```

If the timeout completes, we know that every element at or left of the timeout pointer are timed out. Hence, we can remove these elements with a single operation. We just set head to the element right of the timeout pointer:

```
    tail
     |
head |
 |   |
 V   V
[a,b,c]
```

If the timeout points to the head, and this element is removed, the timer is cancelled. However, we no longer know how many drifts are to be included in the next batch. ~~We now have to do a linear search for the rightmost update that was added in the same bulk (has the same timestamp). Note that we would also have to use linear time to supply the timeout error handlers, so this should not be an issue.~~ [EDIT: That was a stupid idea, we can simply keep track of the tail of each batch that was added to the queue. This is referenced from each node, so we can find it immediately. This way we still have constant time for removing the drift that the timeout pointer references. Then we can remove `n` promises in constant time, but note that we still need linear time to supply all of them to a timeout handler in case one is set.
Moreover, referencing the tail of the bulk is done indirectly so we can update it in constant time, too. This is necessary for when the drift decays that all elements of the same bulk operation point to.]

Here, the timeout pointer followed `a`'s reference to `c` because `c` was the last element that was added together with `a`.

```
tail   timeout pointer
    \ /
     V
head |
 |   |
 V   V
[a,b,c]
 | | ^
/–/–/ \
|      \
V       |
{ tail: c } <– this is the indirect pointer that keeps track of the bulk tail, it is referenced by a-c
```

It is also possible to add basic backpressure support. When adding elements to the list, we simply return a promise. If the adding operation still leaves room for more drifts in the decaying deque (length < concurrency limit), the returned promise is resolved immediately. If the concurrency limit was reached, we return an unresolved promise. Whenever drifts are decaying (exiting the queue), be it from resolving/rejecting or from timing out, they check whether any adding operations returned pending promises, and resolve them if there is space again.

### Drift datatype

It may make things a bit clearer to specify the following.

```ts
interface Drift<T> {
  prev: Drift<T> | null // null for head drift
  next: Drift<T> | null // null for tail drift
  bulk: { tail: Drift<T> } // it is `bulk.tail === this` iff this drift is the last one in a batch

  elem: Promise<void>
  
  since: number // Date
  source: T
}
```

## Modularization

It is simple to extract the interface of this module, so we can have competing implementations for this. However, I'm not sure if that is really needed. This suggestion is already extremely efficient and improvements can be made via PR. Given that this is a very integral part of the library, I don't think there's too big of a need for customization. Most people just won't care, so why make this a dependency?

## Related work

I did some research on promise queues, and there are already a number of different ones out there. Some are better and some are worse, but I could not find a single one that can offer what a native implementation can achieve. All of them are slower in some sense, or less flexible, or both. None of them was made for our use case.

Also, I did not find anything that weighs less than 30 kB, and most are rather 120 kB.

It would certainly be nice to simply `npm install` something here and roll with it, but I honestly believe that it will cause us more pain in the long term than writing down the above suggestion ourselves in a day or two.

## Conclusion

I'm intrigued what you all think about this, and I'm happy to answer questions.

I am looking forward to implementing this, but I don't know how long it will actually take in the end. Also: help appreciated.

**Reply by KnorpelSenf** [reply]
We may want to drop the optimization that allows us to timeout `n` updates in `O(1)` in case no handler is installed. Dropping this does save a tiny bit of memory and improves some constants.

The reason why the optimization does actually not help that much is because all those promises have to be touched again by V8, no matter what. It could be when completing them, when garbage collecting them, during `free` when shutting down, or whatnot. Also, we advise that a timeout handler is always set, so once a bot is popular enough that performance matters here, it will have a timeout handler installed, and so we must convert the drift chain to an array before passing it (or even invoking the error handler per context object). Either way, going below linear time does not help much.

Dropping this makes it easier to implement this proposal.

**Reply by wojpawlik** [reply]
I've read the first 3 sections, and I have two concerns:
- this would require reimplementing error handling,
- trying to decouple fetching updates from processing them would degrade update confirming, aka syncing offset.

**Reply by KnorpelSenf** [reply]
Thanks for the feedback!

> * this would require reimplementing error handling,

That is true. I see this as a benefit. We finally get proper error handling. I'm not worried about touching the ten lines of code that are responsible for this right now, our error handling consists of not much more than a try-catch with a print statement.

> * trying to decouple fetching updates from processing them would degrade update confirming, aka syncing offset.

I'm not sure what you mean. Can you elaborate what you mean by degrading? It will still be possible to sync the offset, the shutdown behaviour is not necessarily affected (unless we deem it worth revising). One thing about the `getUpdates` call that does indeed change is that we can make the next call _while_ the middleware is still running, hence providing concurrency. This should improve the bot's responsiveness by a lot, but of course it is also something to be aware of.

**Reply by wojpawlik** [reply]
> An update is considered confirmed __as soon as__ `getUpdates` is called with an `offset` higher than its `update_id`.

Confirmed updates aren't delivered to the bot anymore. Updates may be dropped if confirmed too early and bot crashes.

**Reply by KnorpelSenf** [reply]
That is true. That's also how the current codebase behaves. If you receive an update and your middleware takes forever, the timeout kicks in, then the next update is requested.

It wasn't much better before #1224—the timeout case was not handled at all and the bot just got stuck. At least it didn't skip any updates, but that behaviour also did not help processing those updates in any meaningful way, either.

It's good you point this out, though! Thanks! With the introduction of the described error handling, we are actually able to delay the `getUpdates` call until all middleware either completed successfully (resolve), or was handled appropriately (reject or timeout). This is easily done using the simple backpressure mechanism that I wrote about.

We should probably value this over the improved concurrency we could achieve, and only let it be switched on optionally.

**Reply by wojpawlik** [reply]
> ```ts
> prev: Drift<T> | null // null for head drift
> ```

Without this, it's tricky to remove an element form the decaying queue, but it's not used for anything else, I don't like it... this would be unnecessary if queue items are simply marked as done instead of removed immiediately.

**Reply by wojpawlik** [reply]
How does double linking affect garbage collection?

**Reply by KnorpelSenf** [reply]
> I don't like it

If you want to be able to delete an element from a collection in constant time, having a _doubly_ linked list is pretty much the only way to go (in practise). You cannot avoid the backlinks. (We could make them implicit by capturing this in a closure, but that's just making things worse.)

> How does double linking affect garbage collection?

Great question, that is interesting stuff. In short: the garbage collector has a hard time to detect large abandoned (cyclic) structures. This is an expensive operation. In general, it is the most efficient to decompose your object graph before unreferencing its nodes.

It is very common when implementing doubly linked lists to unlink all elements before removing parts of the chain. This is more efficient that removing the subchain in a single operation and then confronting the GC with graph component detection.

Given that we only remove single elements, this is not an issue. However, when we perform the timeout, it is definitely best to iterate through the removed sublist and remove all links between the elements. (Note that we have to loop through them anyway because we supply them to the timeout handler.)

**Reply by wojpawlik** [reply]
> we have to loop through them anyway because we supply them to the timeout handler.

Unless we pass iterable to timeout handler.

> However, when we perform the timeout, it is definitely best to iterate through the removed sublist and remove all links between the elements. 

Alternatively, we could leave the list as-is, and leave it to handler timeout to clear it (probably by aborting). Doing otherwise is kinda ignoring the problem: bot still cannot be stopped gracefully because middleware haven't finished.

**Reply by KnorpelSenf** [reply]
There is no unless here. We have to loop through them no matter what, either we do it or the GC does it more expensively.

> > However, when we perform the timeout, it is definitely best to iterate through the removed sublist and remove all links between the elements.
> 
> Alternatively, we could leave the list as-is, and leave it to handler timeout to clear it (probably by aborting). Doing otherwise is kinda ignoring the problem: bot still cannot be stopped gracefully because middleware haven't finished.

Not quite sure what you mean by that, I feel like we mean the same thing here.

**Reply by wojpawlik** [reply]
You seem to imply that when timeout handler returns, Telegraf should clear the list. I'm wondering if the list shouldn't be left alone, with the expectation that if handler timeout doesn't throw, middlewares will be aborted or something and finally settle.

**Reply by KnorpelSenf** [reply]
So
- timeout handler throws: remove from list
- else: assume that the promise will resolve eventually, simply wait for it without safety net.

Correct?

**Reply by wojpawlik** [reply]
Timeout handler throws: crash (this should be the default handler).

> * else: assume that the promise will resolve eventually, simply wait for it without safety net.

Correct. It's more correct, and simpler (no worrying about GC). We'll see how it goes and perhaps change later.

**Reply by KnorpelSenf** [reply]
> Timeout handler throws: crash (this should be the default handler).

By crashing you mean to throw an error in a promise that cannot be handled, right? So crash for Node.js > v15 or something, and print an error for other versions.

I'm fine with the other behaviour.

**Comment by KnorpelSenf** [comment]
We could even implement partial sorting, i.e. guarantee that updates from the same chat/user/whatever are processed sequentially. This is something we were never able to do before because we did not keep track of when each individual update was done processing!

**Reply by wojpawlik** [reply]
How? Also, it should be optional: `@rextester_bot` wants to see and respond to latest message edits ASAP.

**Reply by KnorpelSenf** [reply]
I had a good idea that I forgot by now. The naive approach is to store a map in addition to the list, which is really ugly. Double the memory, access times are not changed much, though.

I did research about the memory thing I was concerned about, but it seems like heavy writes to maps are not an issue.

I believe we should think about this last, if at all.

**Reply by wojpawlik** [reply]
There's one relatively good `Map` approach:

```ts
const dummy = Promise.resolve()

const serialize = (): Middleware.Fn<Context> => {
  const map = new Map<number, Promise<unknown>>()
  return (ctx, next) => {
    if (ctx.chat === undefined) return next()
    const oldPromise = map.get(ctx.chat.id) ?? dummy
    const newPromise = oldPromise.finally(next)
    map.set(ctx.chat.id, newPromise)
    return newPromise
  }
}
```

It's completely orthogonal to this discussion, so I agree we shouldn't think about it here.

**Reply by wojpawlik** [reply]
Also, creator of `telegraf-throttler` told me that's it's also meant to `serialize` incoming updates by person.

**Comment by wojpawlik** [comment]
Before diving into https://github.com/telegraf/telegraf/discussions/1234#discussioncomment-202146, I want to add that crashing is not actually as bad as I made it sound initially. When running under some process manager, it's equivalent to reporting the problem, discarding the problematic updates and restarting the process.

I'm not against allowing users to customize this behavior, but I doubt many would. Code using promises cannot be aborted by default.

`Telegraf::handleUpdate` could maintain a `Set` of currently processed updates, to identify which one(s) hanged the bot.

As long as the interface and sensible confirming are maintained, switching to a more sophisticated mechanism would be non-breaking.

Worst part is, this naive approach would make it impossible to solve #1207 I believe.

**Reply by KnorpelSenf** [reply]
> I'm not against allowing users to customize this behavior, but I doubt many would.

I disagree that this is what it's about. I'm fine with no one else customizing it, I'm willing to implement it. Already for myself, I have to have support for this, because I want to have control over the software I write. For instance, having uncaught promise rejections in any case is not acceptable. Being unreliable by design is a reason to either fork or ditch the lib. (I'm literally pausing any bot development right now because `telegraf` is sometimes so unpredictable … reasons why I'm here)

> Code using promises cannot be aborted by default.

That is both true and not what this proposal is about, so I'm not quite sure how this fits in.

> `Telegraf::handleUpdate` could maintain a `Set` of currently processed updates

It's a good idea. My suggestion solves the same problem in a similar way, but it does this more efficiently, more transparently, and it allows more features to be built on top.

> switching to a more sophisticated mechanism would be non-breaking

I'm still not so sure about this. Changing the way errors are handled can affect in which cases the process exits. It might be that this will in fact not be the case for us—we just don't know enough about the implementation details (yet) in order to guarantee this, that's why I'm so hesitant to agree.

**Reply by wojpawlik** [reply]
> but it does this more efficiently

Because it has _unamortized_ `O(1)` push and pop? It also maintains order, for extra determinism.

The main point of the queue (or set) is to be able to revisit updates to see if they timed out, correct? This would be done periodically?

**Reply by KnorpelSenf** [reply]
> > but it does this more efficiently
> 
> Because it has _unamortized_ `O(1)` push and pop? It also maintains order, for extra determinism.

Exactly. Also, hashes are not too memory-efficient. I don't know too much about how V8 handles them, but having _exactly the necessary memory_ with constant access time is unlikely to be beaten by a regular hash map (which is what most sets use internally). Depending on V8 magic, constant growing and shrinking of the number of elements in the set are may also bloat memory consumption, while the described linked list will be garbage collected after few every decay operations, so the memory consumption scales down again when the load on the bot decreases. Sets (usually) don't do that.

> The main point of the queue (or set) is to be able to revisit updates to see if they timed out, correct?

Yes and no. They are only revisited as soon as they time out. So yes, they are revisited when they time out, and no, we don't need to check this. They are not visited when they are not timed out yet. We may want to inspect in some cases _how long it will take_ until a certain update times out, but this will not be done more often than every `handlerTimeout` or per purge, and always only for a single udpate that represents that batch. Even this update is determined in constant time, we don't need to search for it or something.

As we preserve ordering, we are able to supply a list of of timed out updates to the error handles (or call it multiple times in the correct order with one update each, if you want).

> This would be done periodically?

No. We know the time when the drifts were added, so we just set a timer that waits exactly the right amount of time and then purges exactly the right elements, as they are known in advance.

**Reply by KnorpelSenf** [reply]
Maybe I should be clear about the sequence of operations.

1) Update comes in (let's assume it is the only one in its batch for simplicity)
1) Update is converted to drift, with current timestamp
1) Drift is added to the queue at tail
1) Timer is started with `handlerTimeout`
1) Let's assume the drift does not decay in time, then it must be referenced by head
    - The timeout kicks in
   -  It removes the head drift and supplies it to the error handler
   - It inspects the timestamp of the new head
   - It starts again 
1) Let's assume that the drift does decay
   - The drift decays and removes itself
   - We know the timer is referencing it (head is pointing at it), so we can
      - stop the timer
      - inspect the new head
      - start the timer for that drift again

**Comment by KnorpelSenf** [comment]
I will begin implementing this tomorrow if there are no further concerns. https://github.com/telegraf/telegraf/discussions/1234#discussioncomment-203173 already helped a lot. We might find more during testing

**Comment by wojpawlik** [comment]
Wouldn't it be simpler to have initial implementation store drifts in an array (which could be created using `Array.prototype.map`) and scan thru it after `handlerTimeout` in `O(n)`?

**Reply by KnorpelSenf** [reply]
Very good point. I thought about this, too, and it would indeed be much simpler to implement. The problem is that we cannot know whether a promise is pending or not.

My initial suggestion would've been something like that, but it does not help us to loop through the array if we cannot determine which elements are resolved. That means we cannot distinguish between those elements that actually completed successfully, and those that timed out. Consequently, while we could simply remove all old promises, we have no idea where to handle a timeout error and where not.

The only way to find out when a promise completes is by calling both `then` and `catch` on it. This is precisely why we cannot add the promises to the deque just like that, but instead have to “transform” them before, i.e. `p => p.then(..).catch(..)`. The code in these `then` and `catch` callbacks could simply set a status flag for the respective element, but I figured it would be much easier to simply remove the promise from the queue right away. In a doubly linked list, this is trivial (no index shifting).

This induces a special behaviour of a deque with self-removing promises, and I coined the term _drift_ to capture this concept.

**Reply by wojpawlik** [reply]
We could store a `done` flag in each drift, and set it in `finally`

**Reply by KnorpelSenf** [reply]
That is true, that's what I meant with status flag, sorry for not being clear. Still it's not much easier to implement then. We must track the status of all elements and run a periodic timer which also must have a proper lifecycle. Remeber that a running timer keeps the Node.js process alive, so running the timer every two seconds means that the shutdown will potentially be delayed that long, i.e. we have to cancel it immediately and start it again depending on whether all flags in the list are set to `done` and so on and so forth. I don't feel well with running it every 50 ms, though.

Implementing a linked list really is concise, especially when we only need the two operations add and decay :)

**Reply by wojpawlik** [reply]
https://nodejs.org/api/timers.html#timers_timeout_unref

**Reply by wojpawlik** [reply]
On 2nd thought, using an array is not simple at all. We need a collection of contexts to be able to report which updates timed out, and to be able to communicate with middleware (probably to tell it to abort).

(Speaking of aborting, all `ctx.tg` in batch could share one `signal`)

However, `handleUpdate` effectively returns a `Promise<void>`, so it would need to be split into creating contexts and processing them... No such issue with a linked list, to which `handleUpdate` can simply append at any time.

I wouldn't reimplement a linked list, https://yarnpkg.com/package/yallist has everything needed: `.pushNode`, `.removeNode`, `.length`, `[Symbol.iterator]`, typings, no dependencies.

**Reply by KnorpelSenf** [reply]
> using an array is not simple at all

I'd say it's about as simple as linked lists, but just way less elegant with the periodic timer and copying the array all the time. An array is not the right data structure for this problem.

> (Speaking of aborting, all `ctx.tg` in batch could share one `signal`)

I agree.

> I wouldn't reimplement a linked list, https://yarnpkg.com/package/yallist has everything needed: `.pushNode`, `.removeNode`, `.length`, `[Symbol.iterator]`, typings, no dependencies.

Too late.

Also, we do in fact have a few minor benefits from adding our own 10 lines of linked list implementation, mainly that we can integrate it very neatly with the timer handling, but also faster startup time (no extra dependency to parse).

**Reply by wojpawlik** [reply]
List should live in `core/helpers`, so it's an extra file to parse either way.

**Comment by wojpawlik** [comment]
No matter what we do, `handlerTimeout` should be longer than fetch timeout, which should be longer than `getUpdates` timeout.

`getUpdates` has a max timeout of 50s, so I suggest 60s for fetch timeout and 90s for `handlerTimeout`.

I don't mind crashing by default after such a long timeout at all.

**Reply by KnorpelSenf** [reply]
> fetch timeout, which should be longer than `getUpdates` timeout

That makes a lot of sense.

> `handlerTimeout` should be longer than fetch timeout

That does not. It is a legit use case to only resend the `getUpdates` request every 10 min if no updates are coming in, while still demanding that the middleware completes in seconds once an update arrives. In this case it would be `handlerTimeout` <<< fetch timeout.

> `getUpdates` has a max timeout of 50s, so I suggest 60s for fetch timeout and 90s for `handlerTimeout`.

Nonetheless I'm fine with that, those defaults sound good.

> I don't mind crashing by default after such a long timeout at all.

Which of the three timeouts are you referring to? And why would `telegraf` decide to crash the Node.js process instead of giving control back to the library consumer, other than in a default handler that we advise to overwrite?

**Reply by wojpawlik** [reply]
I'm referring to `handlerTimeout`. Crash would only be default handler.

> It is a legit use case to only resend the `getUpdates` request every 10 min if no updates are coming in, while still demanding that the middleware completes in seconds once an update arrives.

Currently `getUpdates` and other methods share the same hardcoded timeout. It might make sense to make it configurable in  the future.

**Reply by KnorpelSenf** [reply]
Agree with both points :+1:

**Comment by wojpawlik** [comment, marked as answer]
We went for the simplest solution, throw after `handlerTimeout` (#1262), in the future we may implement more sophisticated alternative to `telegraf.ts` (https://github.com/telegraf/telegraf/pull/1276#issuecomment-755753140).
